Add process information to opennid_signing.md (#182)
sigstore/docs: content/en/cosign/openid_signing.md

Currently, Sigstore supports Microsoft, Google, and GitHub. As an alternative, you can also use the environment variable `SIGSTORE_ID_TOKEN` to identify yourself by setting its value to that of the identity token. Cosign also has support for detecting some of these automated environments and producing an identity token. Currently this supports Google and GitHub, in addition to other environments. See [Cosign's providers](https://github.com/sigstore/cosign/tree/main/pkg/providers) for a complete list.

### The signing, witnessing, and verifying process

> Identity-based signing is the default because managing and distributing keys is hard to do securely, and developers are very likely to already have a trusted identity rather than a signing key. When using Sigstore's defaults for signing and verification, the process of signing is as follows:

Reviewer: I think this can be a little tentative, and I would suggest "challenging" or "difficult" over "hard"

```suggestion
Identity-based signing is the default because managing and distributing keys can be challenging to do securely, and developers are very likely to already have a trusted identity rather than a signing key. When using Sigstore's defaults for signing and verification, the process of signing is as follows:
```

#### Verifying identity and signing the artifact

1) An in-memory public/private keypair is created. 
2) The identity token is retrieved.

> 3) Sigstore's certificate authority checks the identity of the user signing the artifact and issues a certificate attesting to their identity. The identity is bound to the public key. Decrypting with the public key will prove the identity of the private keyholder. 

Author: I removed the sentence.

4) For security, the private key is destroyed shortly after and the short-lived identity certificate expires. Users who wish to verify the software will use the transparency log entry, rather than relying on the signer to safely store and manage the private key.

> #### Witnessing and recording the process

Author: changed as requested.

> To create the transparency log entry, a Sigstore client creates an object containing information that will allow signature verification without the (destroyed) private key. The object contains the hash of the artifact, the public key, and the signature. Crucially, this object is timestamped. The Rekor transparency log "witnesses" the signing event by entering a timestamped entry into the records that attests that the secure signing process has occurred. Clients upload signing events to the transparency log so that the events are publicly auditable. Artifact owners should monitor the log for their identity to verify each occurrence.The software creator publishes the timestamped object, including the hash of the artifact, public key, and signature.

Reviewer: Just a space typo here

```suggestion
To create the transparency log entry, a Sigstore client creates an object containing information that will allow signature verification without the (destroyed) private key. The object contains the hash of the artifact, the public key, and the signature. Crucially, this object is timestamped. The Rekor transparency log "witnesses" the signing event by entering a timestamped entry into the records that attests that the secure signing process has occurred. Clients upload signing events to the transparency log so that the events are publicly auditable. Artifact owners should monitor the log for their identity to verify each occurrence. The software creator publishes the timestamped object, including the hash of the artifact, public key, and signature.
```

#### Verifying the signed artifact

> When a software consumer wants to verify the software’s signature, sigstore compares a tuple of signature, key/certificate, and artifact from the timestamped object against the timestamped Rekor entry. If they match, it confirms that the signature is valid because the user knows that the expected software creator, whose identity was certified at the moment of signing, published the software artifact in their possession. The entry in the Rekor’s immutable transparency log means that there’s no need to rely on a potentially insecure private key to perform the verification. 

Reviewer 2: ```suggestion
When a software consumer wants to verify the software’s signature, Sigstore compares a tuple of signature, key/certificate, and artifact from the timestamped object against the timestamped Rekor entry. If they match, it confirms that the signature is valid because the user knows that the expected software creator, whose identity was certified at the moment of signing, published the software artifact in their possession. The entry in the Rekor’s immutable transparency log means that the signer will be monitoring the log for occurrences of their identity and will know if there is an unexpected signing event. 
```

Reviewer 2: Removing mentions of insecure keys, because rekor does support uploading with keys. Using keys isnt inheritently insecure, its just harder to manage.

Currently, Sigstore supports Microsoft, Google, and GitHub. As an alternative, you can also use the environment variable `SIGSTORE_ID_TOKEN` to identify yourself by setting its value to that of the identity token. Cosign also has support for detecting some of these automated environments and producing an identity token. Currently this supports Google and GitHub, in addition to other environments. See [Cosign's providers](https://github.com/sigstore/cosign/tree/main/pkg/providers) for a complete list.

### The signing, witnessing, and verifying process

> When taking advantage of sigstore’s full capabilities, the process of signing is as follows:

Reviewer 2: ```suggestion
When using Sigstore's defaults for signing and verification, the process of signing is as follows:
```

Reviewer 2: Something like this, to call out that this is a default

> 

Reviewer 2: It'd be good to mention that identity-based signing is the default because managing and distributing keys is hard to do securely, and developers are very likely to already have a trusted identity rather than a signing key.

#### Verifying identity and signing the artifact

> 1) A keypair is created. 

Reviewer 2: ```suggestion
1) An in-memory public/private keypair is created. 
```

Reviewer 2: just to be a little more precise

> 2) Sigstore checks the identity of the user signing the artifact and issues a certificate attesting to their identity. 

Reviewer 2: Can we add a step between 1 and 2 to get the identity token?

Author: ok, added a new step saying "The identity token is retrieved."

> 2) Sigstore checks the identity of the user signing the artifact and issues a certificate attesting to their identity. 

Reviewer 2: ```suggestion
2) Sigstore's certificate authority checks the identity of the user signing the artifact and issues a certificate attesting to their identity. 
```

> 3) The certificate is bound to the public key, which adds an extra layer of verification. Exchanging the keys will prove the identity of the private keyholder. 

Reviewer 2: Should 3) be a point under 2)?

Author: i combined 2 and 3.

> 3) The certificate is bound to the public key, which adds an extra layer of verification. Exchanging the keys will prove the identity of the private keyholder. 

Reviewer 2: What do you mean by "exchanging the keys"?

Author: I wasn't sure what this meant either.   Does this mean the public key is somehow exchanged with a symmetric key, or is this sentence simply wrong?

> 3) The certificate is bound to the public key, which adds an extra layer of verification. Exchanging the keys will prove the identity of the private keyholder. 

Reviewer 2: "The certificate is bound to the public key, which adds an extra layer of verification." - This should be rephrased to also include that the identity is bound to the public key, rather than the certificate. Additionally, I would drop the "extra layer of verification".

Author: it now says "The identity is bound to the public key. Decrypting with the public key will prove the identity of the private keyholder. "

> 4) As the artifact is signed, the keys are exchanged, confirming the signer’s identity. 

Reviewer 2: Would drop this line or clarify

5) For security, the private key is destroyed shortly after and the short-lived identity certificate expires. Users who wish to verify the software will use the transparency log entry, rather than relying on the signer to safely store and manage the private key.

#### Witnessing and recording the process

> To create the transparency log entry, Sigstore creates an object containing information that will allow signature verification without the (destroyed) private key. The object contains the hash of the artifact, the public key, and the signature. Crucially, this object is timestamped. The Rekor transparency log "witnesses" the signing event by entering a timestamped entry into the records that attests that the secure signing process has occurred. The software creator publishes the timestamped object, including the hash of the artifact, public key, and signature.

Reviewer 2: ```suggestion
To create the transparency log entry, a Sigstore client creates an object containing information that will allow signature verification without the (destroyed) private key. The object contains the hash of the artifact, the public key, and the signature. Crucially, this object is timestamped. The Rekor transparency log "witnesses" the signing event by entering a timestamped entry into the records that attests that the secure signing process has occurred. The software creator publishes the timestamped object, including the hash of the artifact, public key, and signature.
```

> To create the transparency log entry, Sigstore creates an object containing information that will allow signature verification without the (destroyed) private key. The object contains the hash of the artifact, the public key, and the signature. Crucially, this object is timestamped. The Rekor transparency log "witnesses" the signing event by entering a timestamped entry into the records that attests that the secure signing process has occurred. The software creator publishes the timestamped object, including the hash of the artifact, public key, and signature.

Reviewer 2: Can we also add a line about "Clients upload signing events to the transparency log so that the events are publicly auditable. Artifact owners should monitor the log for their identity to verify each occurrence."

#### Verifying the signed artifact

> When a software consumer wants to verify the software’s signature, sigstore compares the signature from the timestamped object against the timestamped Rekor entry. If they match, it confirms that the signature is valid because the user knows that the expected software creator, whose identity was certified at the moment of signing, published the software artifact in their possession. The entry in the Rekor’s immutable transparency log means that there’s no need to rely on a potentially insecure private key to perform the verification. 

Reviewer 2: Can we clarify that it's comparing not just the signature, but the tuple of (signature, key/certificate, and artifact) matches?

> When a software consumer wants to verify the software’s signature, sigstore compares the signature from the timestamped object against the timestamped Rekor entry. If they match, it confirms that the signature is valid because the user knows that the expected software creator, whose identity was certified at the moment of signing, published the software artifact in their possession. The entry in the Rekor’s immutable transparency log means that there’s no need to rely on a potentially insecure private key to perform the verification. 

Reviewer 2: For the last sentence, let's clarify that the purpose of transparency is to make signing events auditable, not that it lets you not rely on keys. You could still use keys instead of identities and it's still secure and still auditable.
